Implement ConfigurationHandler

gep13 opened the pull request:
To allow for suppressing individual rules.

Relates to #121

gep13 commented:
@tintoy @mholo65 would either of you be able to have a review of this PR?  Thanks!

gep13 commented:
This allows for configuration similar to the following:

```
{
  "chocoLanguageServer":{
    "language": {
      "suppressedRules": [
        "CHOCO0002"
      ]
    }
  }
}
```

**NOTE:** None of this is fixed in stone, just to show how things can be done.

tintoy commented on src/Chocolatey.Language.Server/Handlers/ConfigurationHandler.cs:
```diff
@@ -0,0 +1,157 @@
+            {
+                try
+                {
+                    return (Task<Unit>)OnDidChangeConfiguration(request);
```
At first glance, I'm wondering whether this cast will ever succeed? `OnDidChangeConfiguration` returns `Task.CompletedTask`, not `Task<Unit>`.

gep13 replied:
@tintoy Ah, that one... I think I did that just to get it to compile :smile:  This will likely need some tidy up.

AdmiringWorm commented:
So far it looks good to me

gep13 commented:
@AdmiringWorm @mkevenaar @tintoy I have done a fair bit of refactoring here, but I think it is all for the greater good.

- Refactored so that only one rule per validation
- This allowed adding an Id and DocumentationUrl property to each rule
- Added ability to suppress a rule.  If it is suppressed, the Validate method will never be called on that rule

What are your collective thoughts on these changes?

tintoy commented:
Looks good to me - getting everything from the DI container will keep things simpler :)

gep13 commented:
I am starting to play with this:

![image](https://user-images.githubusercontent.com/1271146/53688760-c7280200-3d40-11e9-9bfd-508940439e5b.png)

Trying to update the configuration, to add suppressions for a particular rule.  Things don't seem to be working though.  Anyone played with this?

mholo65 commented:
@gep13 do you have some code for the above?

gep13 commented:
@mholo65 I have it locally, but I was struggling to get things working the way I want it to. I am thinking about splitting that work into a separate PR.

mkevenaar commented:
Have you looked how https://github.com/DavidAnson/vscode-markdownlint did it? It has a MIT license

gep13 commented:
>@mkevenaar said...
Have you looked how https://github.com/DavidAnson/vscode-markdownlint did it? It has a MIT license

Yes.  I have been using that repository, as well as this one https://github.com/Jason-Rev/vscode-spell-checker, as a reference for what I am trying to achieve.

>@mkevenaar said...
Also I think the CHOCO0### should not be ignorable, as they are a requirement.

I have had the same thought.  While it is correct to say that this is a requirement for pushing to Chocolatey.org, it may not be a requirement when pushing packages internally.  As such, I think I am going to leave the ability to suppress a requirement, but also put in another configuration parameter named something like `allowSuppressionOfRequirements` which will control whether this can be done or not.

AdmiringWorm commented:
> Refactored so that only one rule per validation

I think that makes sense, and I'm all for it.
> This allowed adding an Id and DocumentationUrl property to each rule

Perhaps adding the validation type as a property could also be useful?

> Added ability to suppress a rule. If it is suppressed, the Validate method will never be called on that rule

👍

gep13 commented:
@AdmiringWorm @mkevenaar I have implemented the suggestions that were left in this PR, and I have opened up some follow up issues here:

https://github.com/gep13/chocolatey-vscode/issues/140
https://github.com/gep13/chocolatey-vscode/issues/139

I am going to go ahead and merge this in, as it impacts on a lot of different sections, and I don't want to hold up any further work.  I have tested most of the changes, and it seems to be working fine.  If there are any issues found, we can correct them as we go.

As I mention in the related issue, this will be a breaking change, as I had to refactor the default configuration that is contributed by the extension.

Let me know if you have any questions about anything.

gep13 commented:
@all-contributors please add @tintoy for ideas, review
